store/s3store.go

```diff
@@ -0,0 +1,438 @@
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/sts"
+)
+
+const (
+	MaximumVersions = 100
+	BucketEnvVar    = "CHAMBER_S3_BUCKET"
+
+	latestObjectName = "__latest.json"
+)
+
+// secretObject is the serialized format for storing secrets
+// as an s3 object
+type secretObject struct {
+	Service string                `json:"service"`
+	Key     string                `json:"key"`
+	Values  map[int]secretVersion `json:"values"`
+}
+
+// secretVersion holds all the metadata for a specific version
+// of a secret
+type secretVersion struct {
+	Created   time.Time `json:"created"`
+	CreatedBy string    `json:"created_by"`
+	Version   int       `json:"version"`
+	Value     string    `json:"value"`
+}
+
+// latest is used to keep a single object in s3 with all of the
+// most recent values for the given service's secrets.  Keeping this
+// in a single s3 object allows us to use a single s3 GetObject
+// for ListRaw (and thus chamber exec).
+type latest struct {
+	Latest map[string]string `json:"latest"`
+}
+
+var _ Store = &S3Store{}
+
+type S3Store struct {
+	svc    s3iface.S3API
+	stsSvc *sts.STS
+	bucket string
+}
+
+func NewS3Store(numRetries int) *S3Store {
+	session, region := getSession(numRetries)
+
+	svc := s3.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	stsSvc := sts.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	bucket, ok := os.LookupEnv(BucketEnvVar)
+	if !ok {
+		fmt.Fprintf(os.Stderr, "Must set %s for s3 backend\n", BucketEnvVar)
+		os.Exit(1)
+	}
+
+	return &S3Store{
+		svc:    svc,
+		stsSvc: stsSvc,
+		bucket: bucket,
+	}
+
+}
+
+func (s *S3Store) Write(id SecretId, value string) error {
+	index, err := s.readLatest(id.Service)
+	if err != nil {
+		return err
+	}
+
+	objPath := getObjectPath(id)
```

Comment:
Do we allow/care if people use "/" in their key name? Not sure if there's validation elsewhere, but I think if someone has "/" in the key name, it will make a new folder?

Reply:
We do validation in the CLI code (https://github.com/segmentio/chamber/blob/master/cmd/write.go#L38 for write, but similar for read/etc)